File: meetups/9/matrix.py
# -*- coding: utf-8 -*-

from __future__ import division
import matplotlib.pyplot as plt

import math
import numpy as np
import numpy as np
import matplotlib.cm as cm
import matplotlib.mlab as mlab
import matplotlib.pyplot as plt
import numpy as np
import matplotlib.cm as cm
import  matplotlib.pyplot as plt

step = 10
Lat_st_deg = 30
Lat_st = Lat_st_deg*2*math.pi/360
Lat_f_deg = 42
Lat_f = Lat_f_deg*2*math.pi/360
slope_deg =20
slope = slope_deg * 2*math.pi/360
rot_g_deg = 20
rot_g =rot_g_deg * 2*math.pi/360

input_data = [] # Στον πίνακα Α αποθηκεύονται τα στοιχεία από το input

# Εισαγωγή δεδομένων
fi = open('input')
input = [] # Στον πίνακα input αποθηκεύονται τα raw στοιχεία
counter = 0
while True:
	line = fi.readline()
	if line =='':	
		break
	input.append(line)
	counter = counter + 1
fi.close()
print 'counter = ', counter
matr = np.zeros((counter, 15))
matr_total = np.zeros((19, 37))

# population του πίνακα input_data
for i in input:
	y = [x for x in eval(i)]	
	input_data.append(y)

for i in range(0,len(input_data)):
	for j in range(0,len(input_data[i])):
		matr[i][j] = input_data[i][j]

for i in range(0,(matr.shape)[0]):
	# B (rad)
	matr[i][4] = ((matr[i][0]-81)/364*360)/360*2*math.pi
	# E
	matr[i][5] = 9.87*math.sin(2*matr[i][4])-7.53*math.cos(matr[i][4])-\
	1.5*math.sin(matr[i][4])
	# dt
	matr[i][6] = int(4*(Lat_st_deg-Lat_f_deg)+matr[i][5])
	# solar time
	if matr[i][3]+matr[i][6]<1440:
		matr[i][7] = matr[i][3]+matr[i][6]
	else:
		matr[i][7] = matr[i][3]+matr[i][6] - 1440
	# δ
	matr[i][8] = 23.45*math.sin(math.pi/180*360*(284+matr[i][0])/365)*2*math.pi/360
	# ω
	matr[i][9] = ((matr[i][7]-720)/60*15)*2*math.pi/360

for slope_deg in range(0,95,5):
	slope = slope_deg * 2*math.pi/360
	for rot_g_deg in range(-90,95,5):	
		rot_g =(rot_g_deg-0) * 2*math.pi/360		
		for i in range(0,(matr.shape)[0]):
			matr[i][10] = math.sin(matr[i][8]) * math.sin(Lat_f) * math.cos(slope)-\
			math.sin(matr[i][8]) * math.cos(Lat_f) * math.sin(slope) * math.cos(rot_g)+\
			math.cos(matr[i][8]) * math.cos(Lat_f) * math.cos(slope) * math.cos(matr[i][9])	+\
			math.cos(matr[i][8]) * math.sin(Lat_f) * math.sin(slope) * math.cos(rot_g) * math.cos(matr[i][9])+\
			math.cos(matr[i][8]) * math.sin(slope) * math.sin(rot_g) * math.sin(matr[i][9])	
			# Watt
			if matr[i][10]>0:
				matr[i][11] = matr[i][10]*math.cos(matr[i][9])*matr[i][1]
			else:
				matr[i][11] = 0
				
			matr_total[int(slope_deg/5)][int(rot_g_deg/5)+18] = matr.T[11].sum()/1000
		print 'slope = ', slope_deg, 'rotation = ', rot_g_deg, 'kWh/year = ', matr.T[11].sum()/1000 

File: meetups/9/matrix.rb
require 'matrix'
require 'cmath'
filename=ARGV.shift
input=File.readlines(filename)

def b day
  2*Math::PI*(day-81)/364.0
end

def e b_value
  9.87*CMath.sin(2*b_value)-7.53*CMath.cos(b_value)-1.5*CMath.sin(b_value)
end

def dt e_value,lat_st_deg=30,lat_f_deg=42
  (4*lat_st_deg-lat_f_deg+e_value).to_i
end

def solar_time minutes,dt
  return minutes+dt if minutes+dt < 1440
  minutes+dt - 1440
end

def delta day
  23.45*2*Math::PI/360*CMath.sin((2*Math::PI/365)*(284+day))
end

def omega solar_time_val
  (2*Math::PI/360)*(solar_time_val-720)/60*15
end

def watt delta_val,omega_val,slope_rad,rotation_rad,lat_f=2*Math::PI*42/360.0
  CMath.sin(delta_val) * CMath.sin(lat_f) * CMath.cos(slope_rad)-CMath.sin(delta_val) * CMath.cos(lat_f) * CMath.sin(slope_rad) * CMath.cos(rotation_rad)+	CMath.cos(delta_val) * CMath.cos(lat_f) * CMath.cos(slope_rad) * CMath.cos(omega_val)	+	CMath.cos(delta_val) * CMath.sin(lat_f) * CMath.sin(slope_rad) * CMath.cos(rotation_rad) * CMath.cos(omega_val)+CMath.cos(delta_val) * CMath.sin(slope_rad) * CMath.sin(rotation_rad) * CMath.sin(omega_val)
end

start=Time.now
matrix=input.map do |line|
  values=line.split(',')
  values.map!{|el| el.strip.to_f }
  b_val=b(values[0])
  e_val=e(b_val)
  dt_val=dt(e_val)
  solar_time_val=solar_time(values[3],dt_val)
  values+[b_val,e_val,dt_val,solar_time_val,delta(values[0]),omega(solar_time_val)]
end
p matrix
threads=[]
(0..90).step(5) do |slope|
  threads<<Thread.new do 
    slope_rad=2*Math::PI*slope/360
    (-90..90).step(5) do |rotation|
      rotation_rad=2*Math::PI*rotation/360
      matrix.each do |m|
         p watt(m[8],m[9],slope_rad,rotation_rad)
      end
    end
  end
end
threads.each{|t| t.join}
p Time.now-start